### PR TITLE
make Dockerfile template more flexible

### PR DESCRIPTION
**Pull request by ruffsl**
Adding snippets and logic for maintaining osrf/docker_images, rework of #72.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
In my opinion it would be good to only integrate functionality into this repo which is also being used by it. Or add functionality which is "generic" and otherwise difficult to provide externally. If specific pieces are only needed by the `osrf/docker_templates` repository it should be possible to keep them there. My rational for that is if they such docker template specific things need to change they don't require a change in this repo and can be updated / maintained independently.

Looking through the current patch quite a bit good be achieved "outside" of this repository and wouldn't strictly need to be in `ros_buildfarm`. I would argue they could be kept in e.g. the `osrf/docker_templates` repository where they are actually being used:

* `template_prefix_path` is already being exposed in order to allow the caller to customize the lookup paths. The added logic in `expand_template_prefix_path` passing a `data` dict and requiring the packages to have a `templates` subfolder seem to be very application specific.
* `create_entrypoint` could easily be called after `create_dockerfile` and could therefore live externally.
* The two snippets `add_generated_comment.Dockerfile.em` and `vcs_import.Dockerfile.em` aren't being used in the `ros_buildfarm` package. They could stay in e.g. the `osrf/docker_templates` repository where they are being used.

On the other hand some other changes could be integrated in this repo:

* The changes in `from_base_image.Dockerfile.em` make sense to make the template more flexible.
* The `verbose` option for `create_dockerfile` could be easily added from the outside (by just printing the content of the generated file) but adding that optional keyword argument would be fine as well (even though it is not being used).

I am not saying the first group of changes has no chance of being added here but currently I don't see a strong enough reason why they should be in this repo rather than where they are actually being used. Maybe you can elaborate why you think those should be added here.

**Comment by ruffsl** [comment]
Ok, I've moved dockerfile/entrypoint/path logic to docker_templates, as well as the two other snippets.
See https://github.com/ruffsl/docker_templates/pull/3

I'd like to avoid the print out of dockerfiles from polluting travis logs that will use this templating functions, so I added an optional  verbose argument to avoid breaking any function calls or expected output, but if the print statement are not needed, I'd suggest omitting them.

File: ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em [hunk 1]
```diff
@@ -1,6 +1,15 @@
+@[if 'base_image' in locals()]@
+FROM @base_image
+@[else]@
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em]: With this additional nesting please add some indentation to the line 5, 7, and 9 to keep the template more readable:

```
@[  if arch in ['i386', 'armhf', 'arm64']]@
@[  else]@
@[  end if]@
```

File: ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em [hunk 2]
```diff
@@ -1,6 +1,15 @@
...
 @# same logic as in builder_check-docker.xml.em
 @[if arch in ['i386', 'armhf', 'arm64']]@
 FROM osrf/@(os_name)_@arch:@os_code_name
 @[else]@
 FROM @os_name:@os_code_name
 @[end if]@
+@[end if]@
+@[if 'maintainer_name' in locals()]@
+@[if maintainer_name]@
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em]: Please these two conditions into a single one, e.g.:

```
@[if vars().get('maintainer_name')]@
```

**Comment by dirk-thomas** [comment]
> Ok, I've moved dockerfile/entrypoint/path logic to docker_templates, as well as the two other snippets.
> See ruffsl/docker_templates#3

:+1:

> I'd like to avoid the print out of dockerfiles from polluting travis logs that will use this templating functions, so I added an optional verbose argument to avoid breaking any function calls or expected output, but if the print statement are not needed, I'd suggest omitting them.

The current behavior on the buildfarm does print the content of the Dockerfile to embed that information into the job output. We shouldn't change that without a reason so the new `verbose` option looks good to me.

I will add a couple of nitpick comment to the changes to the `from_base_image.Dockerfile.em` template.

File: ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em (hunk 1 above)
Comment (ruffsl) [inline ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em]: Done

File: ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em (hunk 2 above)
Comment (ruffsl) [inline ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em]: and Done

**Comment by dirk-thomas** [comment]
And merged... Thanks.

**Pull request merged**